partiql-lang/src/test/kotlin/org/partiql/lang/planner/transforms/PartiQLSchemaInferencerTests.kt

```diff
@@ -1070,6 +1155,895 @@ class PartiQLSchemaInferencerTests {
+                                        constraints = setOf(TupleConstraint.Open(false), TupleConstraint.UniqueAttrs(true))
+                                    ),
+                                ),
+                                contentClosed = true,
+                                constraints = setOf(TupleConstraint.Open(false), TupleConstraint.UniqueAttrs(true))
+                            ),
+                        ),
+                        contentClosed = true,
+                        constraints = setOf(TupleConstraint.Open(false), TupleConstraint.UniqueAttrs(true), TupleConstraint.Ordered)
+                    )
+                )
+            ),
+            SuccessTestCase(
+                name = "EXCLUDE with both duplicates",
+                query = """SELECT * EXCLUDE t."a".b.c
+                    FROM <<
+                        {
+                            'a': {
+                                'B': {
+                                    'c': 0,
+                                    'c': true,
+                                    'd': 'foo'
+                                }
+                            }
+                        }
+                    >> AS t""",
+                expected = BagType(
+                    StructType(
+                        fields = mapOf(
+                            "a" to StructType(
+                                fields = mapOf(
+                                    "B" to StructType(
+                                        fields = mapOf(
+                                            // both "c" removed
+                                            "d" to StaticType.STRING
+                                        ),
+                                        contentClosed = true,
+                                        constraints = setOf(TupleConstraint.Open(false), TupleConstraint.UniqueAttrs(false)) // UniqueAttrs set to false
+                                    ),
+                                ),
+                                contentClosed = true,
+                                constraints = setOf(TupleConstraint.Open(false), TupleConstraint.UniqueAttrs(true))
+                            ),
+                        ),
+                        contentClosed = true,
+                        constraints = setOf(TupleConstraint.Open(false), TupleConstraint.UniqueAttrs(true), TupleConstraint.Ordered)
+                    )
+                )
+            ),
+            ErrorTestCase(
+                name = "invalid exclude collection wildcard",
+                query = """SELECT * EXCLUDE t.a[*]
+                    FROM <<
+                        {
+                            'a': {
+                                'b': {
+                                    'c': 0,
+                                    'd': 'foo'
+                                }
+                            }
+                        }
+                    >> AS t""",
```

Comment:
What happens if you added a similar, but different, test?
```
            ErrorTestCase(
                name = "invalid exclude collection wildcard",
                query = """SELECT * EXCLUDE t.a[*]
                    FROM <<
                        {
                            'a': {
                                'b': {
                                    'c': 0,
                                    'd': 'foo'
                                }
                            }
                        },
                        {
                            'a': [
                                {
                                    'c': 0,
                                    'd': 'foo'
                                }
                            ]
                        }
                    >> AS t""",
```

Since `t.a` is a union of `struct` and `list`?

Reply:
I think we might need to add support for it outright, since `null` is modeled as a distinct StaticType. AKA, very similar query that represents many flows that other SQL implementations might experience:
```sql
SELECT * EXCLUDE t.a[*]
FROM <<
  null,
  {
    'a': [
      {
        'c': 0,
        'd': 'foo'
      }
    ]
  }
>> AS t
```

Especially since everything is nullable by default in SQL. Similar experience:
```sql
-- I know this isn't real syntax, but it's meant for demonstration.
CREATE TABLE t1 (
  a STRUCT (
    c INT,
    d STRING
  )
);

-- A similar syntax that Redshift supports would be
CREATE TABLE t1 (
  a SUPER
);
```

`a` in both queries would be nullable, and therefore the following query would be wrongly typed:
```sql
SELECT * EXCLUDE t.a.*
FROM t1 AS t
```

If I'm not mistaken, the type currently returned would be something like:
```
BagType(
  StructType(c: IntType(), d: StringType())
)
```

Since `t` would be the `unionOf(NULL, StructType( ... ))`. But, I imagine the expected would be:
```
BagType(
  StructType()
)
```

Reply:
I agree w/ the point that `EXCLUDE` over type unions are important after some offline discussion. I still need some clarification on what the desired behavior should be w.r.t. typed unions.

For example, consider 

```
SELECT t
EXCLUDE t.a.b
FROM <<
    {
        'a': 1 {
            'b': 2,
            'c': 3
        }
    },
    {
        'a': NULL
    }
>> AS t
```

Currently the schema is something like:

```
Bag(
    Struct(
        a: unionOf(null, struct(b: int, c: int))
    )
)
```
^ basically ignores the union type for `a`.

Would the schema you propose then be:
```
Bag(
    Struct(
        a: unionOf(null, struct(c: int)) -- excluding `b` from the struct in the union
    )
)
```
?

Reply:
Thanks for confirming. Latest commit https://github.com/partiql/partiql-lang-kotlin/pull/1226/commits/d60d08cd1ab4f80bdce344e51c4983a0760ced7f should cover the union case as well as some other edge cases discussed offline.